HARK/ConsumptionSaving/tests/test_ConsPortfolioFrameModel.py

```diff
@@ -0,0 +1,146 @@
+import HARK.ConsumptionSaving.ConsPortfolioFrameModel as cpfm
+import numpy as np
+import unittest
+
+
+class PortfolioConsumerTypeTestCase(unittest.TestCase):
+    def setUp(self):
+        # Create portfolio choice consumer type
+        self.pcct = cpfm.PortfolioConsumerFrameType()
+        self.pcct.cycles = 0
+
+        # Solve the model under the given parameters
+
+        self.pcct.solve()
+
+class UnitsPortfolioConsumerTypeTestCase(PortfolioConsumerTypeTestCase):
+    def test_simOnePeriod(self):
+
+        self.pcct.T_sim = 30
+        self.pcct.AgentCount = 10
+        self.pcct.track_vars += ['aNrm']
+        self.pcct.initialize_sim()
+
+        self.assertFalse(
+            np.any(self.pcct.shocks['Adjust'])
+        )
+
+        self.pcct.sim_one_period()
+
+        self.assertAlmostEqual(
+            self.pcct.shocks['PermShk'][0],
+            0.95908471
+        )
+
+        self.assertAlmostEqual(
+            self.pcct.shocks['TranShk'][0],
+            1.03172631
+        )
+
+        self.assertAlmostEqual(
+            self.pcct.shocks['Risky'][0],
+            1.52812838
+        )
+
+        self.assertAlmostEqual(
+            self.pcct.state_now['pLvl'][0],
+            self.pcct.state_prev['pLvl'][0] * self.pcct.shocks['PermShk'][0]
+        )
+
+        self.assertTrue(
+            np.any(self.pcct.shocks['Adjust'][0])
+        )
+
+        self.assertAlmostEqual(
+            self.pcct.state_now["mNrm"][0],
+            self.pcct.state_prev["aNrm"][0] \
+            * self.pcct.Rfree / self.pcct.shocks['PermShk'][0] \
+            + self.pcct.shocks['TranShk'][0] \
```

Comment:
I want to note one thing in particular about the tests here.

I found that with the new architecture I was getting a slightly different sampling order for shocks. This is of course fine for the library, but it breaks the extremely brittle tests we've had which check for a specific numerical value.

In many places in this test suite, I've changed to this more resilient and accurate style of test. I'm now testing the relationships between model state/shock/etc. values, rather than particular numerical values.

That will mean that if the code is working properly, some change to the distribution sampling order won't break all the tests.

We might try to do this elsewhere in the library before doing a big random number generation shifting change, such as #1026

Reply:
@sbenthall 

That's great, thanks, you are right that our existing tests are much too finicky and I'm glad to learn that you are robustifying them.  To tell the truth, even 6 decimal places (which is what we have dialed down to from a default of 13) is far more than makes much sense for most of our tests.  3 or 4 digits is probably more than enough.